azurelinuxagent/ga/update.py

```diff
@@ -392,17 +432,48 @@ def _process_goal_state(self, protocol, exthandlers_handler, remote_access_handl
 
         try:
             if incarnation != self.last_incarnation:  # TODO: This check should be based in the etag for the extensions goal state
-                exthandlers_handler.run()  # TODO: The info for the extensions within run() should be based on the extensions goal state
+                if not self._extensions_summary.converged:
+                    message = "A new goal state was received, but not all the extensions in the previous goal state have completed: {0}".format(self._extensions_summary)
+                    logger.warn(message)
+                    add_event(op=WALAEventOperation.GoalState, message=message, is_success=False, log_event=False)
+                    if self._is_initial_goal_state:
+                        self._on_initial_goal_state_completed(self._extensions_summary)
+                self._extensions_summary = ExtensionsSummary()
+                exthandlers_handler.run()
 
             # report status always, even if the goal state did not change
             # do it before processing the remote access, since that operation can take a long time
-            exthandlers_handler.report_ext_handlers_status(incarnation_changed=incarnation != self.last_incarnation)
+            self._report_status(exthandlers_handler, incarnation_changed=incarnation != self.last_incarnation)
 
             if incarnation != self.last_incarnation:
                 remote_access_handler.run()
         finally:
             self.last_incarnation = incarnation
 
+    def _report_status(self, exthandlers_handler, incarnation_changed):
+        vm_status = exthandlers_handler.report_ext_handlers_status(incarnation_changed=incarnation_changed)
+        extensions_summary = ExtensionsSummary(vm_status)
+        if self._extensions_summary != extensions_summary:
+            self._extensions_summary = extensions_summary
+            message = "Extension status: {0}".format(self._extensions_summary)
+            logger.info(message)
+            add_event(op=WALAEventOperation.GoalState, message=message)
+            if self._extensions_summary.converged:
+                message = "All extensions in the goal state have reached a terminal state: {0}".format(extensions_summary)
+                logger.info(message)
+                add_event(op=WALAEventOperation.GoalState, message=message)
+                if self._is_initial_goal_state:
+                    self._on_initial_goal_state_completed(self._extensions_summary)
+
+    def _on_initial_goal_state_completed(self, extensions_summary):
+        fileutil.write_file(self._initial_goal_state_file_path(), ustr(extensions_summary))
+        if conf.get_extensions_enabled():
+            previous = self._goal_state_period
+            self._goal_state_period = conf.get_goal_state_period()
+            if self._goal_state_period != previous:
```

Comment:
why this check matters, regardless we need to switch right?

Reply:
we want to log this only 1 time (the switch always happens, one line above, though from the second goal state forward we are "switching" to the same value)

Reply:
alternatively, the code could have been written as 

```
        if conf.get_extensions_enabled() and conf.get_goal_state_period() != conf.get_initial_goal_state_period():
            self._goal_state_period = conf.get_goal_state_period()
            logger.info("Initial goal state completed, switched the goal state period to {0}", self._goal_state_period)

```

Reply:
This bring me general question, Ideal world it should not be but what if both first goal state period and goal state period values are same then do we want to go through this entire new flow?

Reply:
Not sure I follow you question... in the vast majority of VMs the 2 periods will be the same. On HPC machines they will be different. What is your concern?

Reply:
I thought something else. Makes sense now. NM,

Reply:
ok, i'll rewrite the code as i pointed out above, which seems a little clearer